### PR TITLE
Add `irc_server_name` to differentiate between IRC servers for PMs

### PR DESCRIPTION
**Pull request by llmII**
This has been tested and seems to work properly.

I'm not certain why it was previously set as experimental and warned people *loudly*.

This commit adds a network discriminator. The use for that is so PM's are actually private. If you have 1 Discord server but want to relay to 3 networks, you need to run 3 bots. If the nick `nick1` exists on both network 1 and network 2 when the Discord user replied (prior to this commit) it might have been possible that both network's `nick1` got the PM.

This also replaces the expirementalNotice with a way to introduce the usage of the discriminator when messaging from Discord to IRC and nothing is different as far as the IRC user is concerned so no message introduction.

### CODE REVIEW

**Comment by llmII** [comment]
Question: Should we exit on lack of a network discriminator?

**Comment by qaisjp** [comment]
> I'm not certain why it was previously set as experimental and warned people _loudly_.

This might be a bug. It's supposed to send it to only the Discord person that was messaged, and only once. I got a bug report a few weeks ago that it inadvertently messages everyone on IRC or discord.


 (State on whether a Discord person is messaged is stored in memory.)

File: README.md [hunk 1]
```diff
@@ -68,6 +68,7 @@ The config file is a yaml formatted file with the following fields:
 
 | name                | requires restart | default                                        | optional                     | description                                                                                                                                                              |
 | ------------------- | ---------------- | ---------------------------------------------- | ---------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
+| `discriminator`     | Yes              |                                                | No                           | The network discriminator used for when PM from Discord to IRC nick                                                                                                      |
```
Comment (qaisjp) [inline README.md:71]:
```suggestion
| `discriminator`     | Yes              |                                                | No                           | The network discriminator used for when PMing from Discord to IRC nick                                                                                                   |
```

File: bridge/discord.go [hunk 2]
```diff
@@ -453,6 +457,14 @@ func pmTargetFromContent(content string) (nick, newContent string) {
 	nick = subs[0]
 	newContent = strings.TrimPrefix(subs[1], " ")
 
+	nickParts := strings.Split(nick, "@")
+
+	if nickParts[1] != discriminator {
+		return "", ""
+	}
+
+	nick = nickParts[0]
```
Comment (qaisjp) [inline bridge/discord.go:466]: Will crash if the nick is not valid

File: bridge/irc_connection.go [hunk 3]
```diff
@@ -92,7 +89,9 @@ func (i *ircConnection) experimentalNotice(nick string) {
 
 	if !i.pmNoticed {
 		i.pmNoticed = true
-		_, err := d.ChannelMessageSend(i.pmDiscordChannel, "**Private messaging is still in dev. Proceed with caution.**")
+		_, err := d.ChannelMessageSend(
+			i.pmDiscordChannel,
+			fmt.Sprintf("To reply type: %s@%s, message", nick, i.manager.bridge.Config.Discriminator))
```
Comment (qaisjp) [inline bridge/irc_connection.go:94]:
```suggestion
			fmt.Sprintf("To reply type: `%s@%s, your message here`", nick, i.manager.bridge.Config.Discriminator))
```

**Review by qaisjp — CHANGES_REQUESTED** [review]

File: bridge/discord.go (hunk 2 above)
Comment (llmII) [reply bridge/discord.go:466]: Didn't think of that, I'll fix up the code to use a check on the length of nickParts. Or are you saying the strings.Split call will cause the crash?

Comment (llmII) [reply bridge/discord.go]: Added a  fix.

**Comment by llmII** [comment]
> This might be a bug. It's supposed to send it to only the Discord person that was messaged, and only once. I got a bug report a few weeks ago that it inadvertently messages everyone on IRC or discord (but didn't investigate)
> 
> (State on whether a Discord person is messaged is stored in memory.)

I actually still use that state to only send the introductory message on how to utilize PM from Discord towards IRC. That said say I have 2 nicks (n1, n2), 2 networks (net1, net2), 1 discord server (ds), 1 discord username (du).

If n1 == n2 and both n1 and n2 message du on ds where net1 != net2 - how would the 2 discord relays know whether or not to send the reply from du to n1 or n2 vs since they both have prior sent a message to du they could both see the message du sends and forward it to their respective nick. This assumes n1 is on net1 and n2 is on net2. What would be worse is if n1 and n2's nick are the same but they are not the same person.

**Review by qaisjp — COMMENTED** [review]
Wouldn't this still be noisy when there are multiple bots running? (because it won't know the other ones the other ones would always say `Don't know that is...`)

Maybe it's worth adding multi-network support eventually?

**Comment by llmII** [comment]
> Wouldn't this still be noisy when there are multiple bots running? (because it won't know the other ones the other ones would always say `Don't know that is...`)

True, I failed to think of that - perhaps it should do a "check is this a valid identifier" and if it is and the discriminator part of the id doesn't match fail to reply?

> Maybe it's worth adding multi-network support eventually?

That actually would be sort of possible... just make main.go start multiple instances of the bridge in one process, just would probably need to extend bridge's to have a handle to a list of other bridges by discriminator and the host they connect to.

This might need a bit more thought on how to handle not spamming with "we don't know that person". Both multi-networked and detection like above are worth consideration I believe. I would think multi-network is the correct (albeit slightly more difficult to accomplish) approach.

**Comment by llmII** [comment]
Maybe a config var for "known_discriminators" and it'll check if it matches any of those and drop out if it's a known but not it's own, but if not match, the primary one will reply a generic one and list known discriminators?

**Comment by qaisjp** [comment]
> perhaps it should do a "check is this a valid identifier" and if it is and the discriminator part of the id doesn't match fail to reply?
>
> [...]
>
> Maybe a config var for "known_discriminators" and it'll check if it matches any of those and drop out if it's a known but not it's own, but if not match, the primary one will reply a generic one and list known discriminators?

Instead of a `known_discriminators`, we could just ignore the message if the message is well-formed, but has an unknown discrminator.

There is an unfixable issue (without native multi-discord/irc support), though. If someone types an invalid message, they will get a message from each instance of the bot.

> That actually would be sort of possible... just make main.go start multiple instances of the bridge in one process, just would probably need to extend bridge's to have a handle to a list of other bridges by discriminator and the host they connect to.

This wouldn't be ideal as we'd want to only use one `*discordgo.Session` instead of multiple (to reduce the number of API requests to Discord).

**Comment by llmII** [comment]
> Instead of a `known_discriminators`, we could just ignore the message if the message is well-formed, but has an unknown discrminator.

Definitely the easiest way to handle it - when an unknown discriminator is encountered, quit processing the message.

> There is an unfixable issue (without native multi-discord/irc support), though. If someone types an invalid message, they will get a message from each instance of the bot.

I meant multi-irc, not mutli-discord - the latter would require more thought.

> > That actually would be sort of possible... just make main.go start multiple instances of the bridge in one process, just would probably need to extend bridge's to have a handle to a list of other bridges by discriminator and the host they connect to.
> 
> This wouldn't be ideal as we'd want to only use one `*discordgo.Session` instead of multiple (to reduce the number of API requests to Discord).

True, would need a way to pass the same session into the next started bridge. It would require a bit of thought all around for handling multi-irc support but I do believe using the same session with the same guild would definitely work. The bridge part would also need to know which IRCManager per network for what channels and if the same channels are exposed across 2 networks the bridge would have to push it to both IRCManagers.

What would you consider the ideal way to solve this?

**Comment by llmII** [comment]
This solves the valid identifier unknown discriminator ordeal although we still have the issue where they'll get multiple "unknown" messages for invalid identifiers. That really can't be solved without at least multi-irc support I don't think.

**Comment by llmII** [comment]
> Question: Should we exit on lack of a network discriminator?

Should we have the relay quit if the discriminator is empty (I did make it a required setting with no default...). Just didn't know if that was too invasive.

**Comment by qaisjp** [comment]
> Should we have the relay quit if the discriminator is empty (I did make it a required setting with no default...). Just didn't know if that was too invasive.

Yes please!

**Comment by llmII** [comment]
Changed to exit when discriminator isn't set, also log a message letting one know they need to set it.

**Review by qaisjp — APPROVED** [review]
I have no idea how PMs were working for you but it wasn't working for me without f777d64249a62693fee8640a14e2216845155ac6!

**Comment by llmII** [comment]
Sorry, hadn't gotten around to testing it since you merged wim's discord and I figure that it not working till you fixed it probably had something to do with that.

**Comment by qaisjp** [comment]
Must be that :D The old way was pretty bad anyway

**Pull request merged**